File: functions/src/models/post_response.ts
import { PostModel } from "./post_model";
import { UserModel } from "./user_model";

export class PostResponse{
    post: PostModel;
    user: UserModel;
    constructor({post, user}: {post: PostModel, user: UserModel}){
        this.post = post;
        this.user = user;
    }
    toMap(){
        return {
            'post': this.post.toMap(),
            'user': this.user.toMap()
        }
    }
}

File: lib/app/models/post_response.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'dart:convert';
import 'package:cartisan/app/models/post_model.dart';
import 'package:cartisan/app/models/user_model.dart';

class PostResponse {
  UserModel owner;
  PostModel post;
  PostResponse({
    required this.owner,
    required this.post,
  });

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'owner': owner.toMap(),
      'post': post.toMap(),
    };
  }

  factory PostResponse.fromMap(Map<String, dynamic> map) {
    return PostResponse(
      owner: UserModel.fromMap(
          (map['user'] ?? map['owner']) as Map<String, dynamic>),
      post: PostModel.fromMap(map['post'] as Map<String, dynamic>),
    );
  }

  String toJson() => json.encode(toMap());

  factory PostResponse.fromJson(String source) =>
      PostResponse.fromMap(json.decode(source) as Map<String, dynamic>);
}
